### PR TITLE
monitoring: hide alerts table

### PR DESCRIPTION
**Pull request by michaellzc**
This PR can stop rendering of alerts-related stuff on dashboard when no alert is defined in any one of the `Observable`

## Test plan

Setup kubectl context to dogfood

```sh
kubectl port-forward svc/prometheus 30090
```

```sh
cd monitoring && go generate ./...
```

```sh
sg run grafana
```

CleanShot 2022-05-03 at 15 39 50

### CODE REVIEW

**Comment by michaellzc** [comment]
Current dependencies on/for this PR:
* main
  * **PR #34808** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/34808" target="_blank">Graphite</a> 
    * **PR #34809** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/34809" target="_blank">Graphite</a>  (this PR)

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/34809?utm_source=stack-comment).

**Comment by bobheadxi** [comment]
@michaellzc thoughts on https://github.com/sourcegraph/sourcegraph/pull/34808#issuecomment-1116741617 instead? If there are no alerts here I don't think we should stretch the semantics of `Container` for custom dashboards without doing some wholistic exploration/design work

File: monitoring/monitoring/monitoring.go [hunk 1]
```diff
@@ -41,6 +41,9 @@ type Container struct {
 	// This is used to configure monitoring features that depend on information exported
 	// by the standard Sourcegraph debug server.
 	NoSourcegraphDebugServer bool
+
+	// NoAlertsDefined disables the rendering of alerts table
+	NoAlertsDefined bool
```
Comment (bobheadxi) [inline monitoring/monitoring/monitoring.go:46]: Can this just be inferred? If all observables in all groups have no alerts, disable the alerts table?

**Review by bobheadxi — CHANGES_REQUESTED** [review]
> should I hack into [renderRules](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/monitoring/monitoring/monitoring.go?L286:21) and renderDocumentatino as well so we can get rid of all the observable transformer?

Could you clarify this a bit?

**Comment by michaellzc** [comment]
> > should I hack into [renderRules](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/monitoring/monitoring/monitoring.go?L286:21) and renderDocumentatino as well so we can get rid of all the observable transformer?
> 
> Could you clarify this a bit?

the `NoAlerts` flag right now only prevents the rendering of the alerts toggle and alerts table in the dashboard, but it will still render prometheus rules and the auto-gen alerts documentation unless we add the custom transformer like

```go
containersNoAlertTransformer := func(observable shared.Observable) shared.Observable {
		return observable.WithNoAlerts(`Alerts are enabled in service-specific dashboard.`)
}

shared.NewKubernetesMonitoringGroup(containerName, monitoring.ObservableOwnerDevOps, &shared.KubernetesMonitoringOptions{
  PodsAvailable: containersNoAlertTransformer,
}),
```

Notes: if we can infer this from the groups, we shouldn't need to do anything special to the current generator for docs and prom alerts rules

> Can this just be inferred? If all observables in all groups have no alerts, disable the alerts table?

That would be great and I will take a look

File: monitoring/monitoring/monitoring.go [hunk 2]
```diff
@@ -79,29 +93,33 @@ func (c *Container) renderDashboard() *sdk.Board {
...
+			// Show alerts matching the selected alert_level (see template variable above)
+			Expr:        fmt.Sprintf(`ALERTS{service_name=%q,level=~"$alert_level",alertstate="firing"}`, c.Name),
+			Step:        "60s",
+			TitleFormat: "{{ description }} ({{ name }})",
+			TagKeys:     "level,owner",
+			IconColor:   "rgba(255, 96, 96, 1)",
+			Enable:      false, // disable by default for now
+			Type:        "tags",
+		}}
+	}
```
Comment (michaellzc) [inline monitoring/monitoring/monitoring.go:122]: do you prefer to merge these two blocks under the same if condition? 
does it mater whether we're mutating `board.Annotations.List` or `board.Templating.List` first?

File: monitoring/monitoring/monitoring.go [hunk 3]
```diff
@@ -66,6 +66,20 @@ func (c *Container) validate() error {
 	return nil
 }
 
+// isAlertDefined indicates if a dashboard has any alert defined.
+func (c *Container) isAlertDefined() bool {
```
Comment (bobheadxi) [inline monitoring/monitoring/monitoring.go:70]: Alerts should be defined as a preferred default, so this condition should be inverted:

```suggestion
func (c *Container) noAlertsDefined() bool {
```

File: monitoring/monitoring/monitoring.go (hunk 2 above)
Comment (bobheadxi) [reply monitoring/monitoring/monitoring.go:122]: I think the ordering of variables inserted into annotations and templating matters, if that can be preserved I don't mind the blocks being joined

File: monitoring/monitoring/monitoring.go [hunk 4]
```diff
@@ -79,29 +93,33 @@ func (c *Container) renderDashboard() *sdk.Board {
...
 	board.AddTags("builtin")
-	alertLevelVariable := ContainerVariable{
-		Label: "Alert level",
-		Name:  "alert_level",
-		Options: ContainerVariableOptions{
-			Options: []string{"critical", "warning"},
-		},
-	}
-	board.Templating.List = []sdk.TemplateVar{alertLevelVariable.toGrafanaTemplateVar()}
+	if !c.noAlertsDefined() {
```
Comment (caugustus-sourcegraph) [inline monitoring/monitoring/monitoring.go:96]: I'm sympathetic to @bobheadxi's [point here](https://github.com/sourcegraph/sourcegraph/pull/34809#discussion_r865159551) about alerts being the default, but I do want to mention that `!c.noAlertsDefined` is harder to think about than `c.isAlertDefined`.

**Review by caugustus-sourcegraph — APPROVED** [review]

File: monitoring/monitoring/monitoring.go (hunk 4 above)
Comment (bobheadxi) [reply monitoring/monitoring/monitoring.go:96]: I'd be fined with a reassignment a la `hasAlertsDefined := !c.noAlertsDefined()` and using that throughout, but on the `Container` itself we should export an API that is representative of the default state

**Comment by michaellzc** [comment]
**Graphite Merge Job**

Current status: 🔄 Waiting for checks to pass

This pull request is currently waiting for CI and mergeability checks to pass before being merged.

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/34809?utm_source=stack-comment).


Job Reference: RicP9zqDgyRrH5qyJYZs

**Pull request merged**